Skip headers from jupyter kernelspec list.

ydl7 opened the pull request:
On certain configurations, `juypter kernelspec list` has a header with irrelevant information. This causes an error when parsing JSON output and can simply be ignored.

Example header:

```
0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
```

ydl7 committed:
Skip headers from jupyter kernelspec list.

nnicandro commented:
Do you still get those messages even when applying the fix in #418?

lesteve commented:
I get this warning with Python 3.11. The fix in #418 (adding `--log-level ERROR`) does not help:
```
❯ jupyter kernelspec list --json --log-level ERROR
0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
{
  "kernelspecs": {
    "python3": {
      "resource_dir": "/home/lesteve/.local/share/jupyter/kernels/python3",
      "spec": {
        "argv": [
          "python",
          "-m",
          "ipykernel_launcher",
          "-f",
          "{connection_file}"
        ],
        "env": {},
        "display_name": "Python 3 (ipykernel)",
        "language": "python",
        "interrupt_mode": "signal",
        "metadata": {
          "debugger": true
        }
      }
    }
  }
}
```

It seems to me that the best fix is to only consider stdout when executing jupyter commands as was proposed in #388, rather than adding more and more logic to discard unuseful info, which is what #418 did and and this PR proposes.

alberti42 committed:
jupyter-monads.el: Move state-helper defuns below their macros

The state-helper defuns added in a0871ba (jupyter-get-client,
jupyter-push, jupyter-pop, jupyter-set-client, jupyter-at-point) appear
in the file *before* the jupyter-mlet* and jupyter-do macros they
expand to.  When the file is byte-compiled, the macros are not yet
known at the point those defuns are processed, so the calls to
jupyter-mlet* / jupyter-do are emitted as plain function calls instead
of being macro-expanded inline.

The compiler signals 'macro `jupyter-do' defined too late' (and an
analogous condition for jupyter-mlet*).  At runtime, calling e.g.
jupyter-get-client signals (void-variable state): the body
\`(let ((client (if (listp state) ...))) ...)\` is evaluated in the
defun's scope, where state is unbound, before being passed as an
argument to the function-call form of jupyter-mlet*.

This breaks jupyter-repl-sync-execution-state (called from
jupyter-repl-mode), which in turn aborts jupyter-bootstrap-repl before
its :after method runs, leaving freshly-associated buffers without a
client.

Reorder so the macro definitions precede their callers.  No behavioral
change in interpreted mode; in compiled mode the macros now expand
inline as intended.

* jupyter-monads.el (jupyter-get-client, jupyter-push, jupyter-pop)
(jupyter-set-client, jupyter-at-point): Move below jupyter-mlet*,
jupyter-do, and jupyter-run-with-state.
